pennylane/gradients/gradient_transform.py

```diff
@@ -114,70 +114,94 @@ def assert_no_tape_batching(tape, transform_name):
 
-            if not qml.operation.has_grad_method(op):
-                # no differentiation method is registered for this operation
-                info["grad_method"] = None
 
-            elif (tape._graph is not None) or use_graph:
-                # if `op` is an observable, we want the associated MP that's in the graph
-                op_or_mp = tape[par_info["op_idx"]]
-                if not any(tape.graph.has_path(op_or_mp, mp) for mp in tape.measurements):
-                    # there is no influence of this operation on any of the observables
-                    info["grad_method"] = "0"
-                    continue
+def _find_gradient_methods(tape, trainable_param_indices, use_graph=True):
+    """Returns a dictionary with gradient information of each trainable parameter."""
 
-            info["grad_method"] = op.grad_method
+    return {
+        idx: _try_zero_grad_from_graph_or_get_grad_method(
+            tape, tape.trainable_params[idx], use_graph
+        )
+        for idx in trainable_param_indices
+    }
 
 
-def _grad_method_validation(method, tape):
+def _validate_gradient_methods(tape, method, diff_methods):
     """Validates if the gradient method requested is supported by the trainable
     parameters of a tape, and returns the allowed parameter gradient methods."""
-    diff_methods = {
-        idx: info["grad_method"]
-        for idx, info in enumerate(tape._par_info)  # pylint: disable=protected-access
-        if idx in tape.trainable_params
-    }
 
     # check and raise an error if any parameters are non-differentiable
-    if nondiff_params := {idx for idx, g in diff_methods.items() if g is None}:
+    nondiff_params = [tape.trainable_params[idx] for idx, m in diff_methods.items() if m is None]
+    if nondiff_params:
         raise ValueError(f"Cannot differentiate with respect to parameter(s) {nondiff_params}")
 
-    numeric_params = {idx for idx, g in diff_methods.items() if g == "F"}
-
-    # If explicitly using analytic mode, ensure that all parameters
-    # support analytic differentiation.
+    # If explicitly using analytic mode, ensure that all
+    # parameters support analytic differentiation.
+    numeric_params = [tape.trainable_params[idx] for idx, m in diff_methods.items() if m == "F"]
     if method == "analytic" and numeric_params:
         raise ValueError(
             f"The analytic gradient method cannot be used with the parameter(s) {numeric_params}."
         )
 
-    return tuple(diff_methods.values())
-
 
-def gradient_analysis_and_validation(tape, method, use_graph=True, grad_fn=None, overwrite=True):
-    """Update the parameter information dictionary of the tape with gradient information of
-    each parameter. Then validates if the gradient method requested is supported by the trainable
-    parameters of a tape, and returns the allowed parameter gradient methods.
+def find_and_validate_gradient_methods(tape, method, parameters, use_graph=True):

```

Comment:
Since when I see `parameters`, I would think something like `(np.array(0.5), np.eye(2), ...)`, maybe `trainable_param_indices`, or `param_idxs`, or something.